### PR TITLE
Add Triton benchmark support in compare script

### PR DESCRIPTION
**Pull request by ESI-SYD**

### CODE REVIEW

**Comment by vlad-penkin** [comment]
@ESI-SYD please do not merge this PR without my approval. 

I'm working on the completely new script for comparing the E2E accuracy results. All current script code for E2E performance results comparisons most likely will become OBE after it.

File: scripts/compare-ci-runs/compare_runs.py [hunk 1]
```diff
@@ -188,12 +203,40 @@ def eval_data(df: pd.DataFrame, numerator: str, denominator: str, plot: bool):
...
             ax = sns.boxplot(df, x="xlabel", y="relative difference", order=order)
 
-            ax.set(xlabel=None, ylabel="Relative difference in speedup")
+            ax.set(xlabel=None, ylabel="Relative difference in performance")
 
             pdf.savefig(fig, bbox_inches="tight")
             print(f"Saved performance plot to {filename}")
 
 
+def eval_data(args: argparse.Namespace, df: pd.DataFrame, numerator: str, denominator: str):
```
[the remark this answers is not in the document]
Comment (ESI-SYD) [reply scripts/compare-ci-runs/compare_runs.py]: Refined.

File: scripts/compare-ci-runs/compare_runs.py [hunk 2]
```diff
@@ -74,48 +74,63 @@ def get_raw_data(args: argparse.Namespace) -> tuple[Optional[Path], Optional[Pat
...
 
-    raw_data["suite"] = suite
-    raw_data["datatype"] = datatype
-    raw_data["mode"] = mode
-    raw_data.rename(columns={"speedup": f"speedup {config}"}, inplace=True)
+        raw_data["suite"] = suite
+        raw_data["datatype"] = datatype
+        raw_data["mode"] = mode
+        raw_data.rename(columns={"speedup": f"speedup {config}"}, inplace=True)
+    else:
```
[the remark this answers is not in the document]
Comment (ESI-SYD) [reply scripts/compare-ci-runs/compare_runs.py]: Sounds good.

File: scripts/compare-ci-runs/compare_runs.py [hunk 3]
```diff
@@ -142,13 +157,13 @@ def eval_data(df: pd.DataFrame, numerator: str, denominator: str, plot: bool):
...
 
-    print("Overview of relative difference in speedup.\n"
+    print("Overview of relative difference in perf.\n"
           "Relative difference 0.0 means both perform identically,"
           f" relative difference > 0.0 means {numerator} performs better,"
           f" relative difference < 0.0 means {denominator} performs better")
 
     print(df["relative difference"].describe())
-    print(f"Mean speedup for denominator: {df[denom_col].mean()}")
+    print(f"Mean perf for denominator: {df[denom_col].mean()}")
```
[the remark this answers is not in the document]
Comment (ESI-SYD) [reply scripts/compare-ci-runs/compare_runs.py]: Ratio(triton2xetla) in Triton benchmark, added a parameter `perf_index` to handle different scenarios. Does it looks better ?

**Comment by pbchekin** [comment]
@ESI-SYD I've tried the script and it didn't work for me:
```
./compare-ci-runs/compare_runs.py --triton-benchmark -n dev:10752862880 -d ref:10752866039
```
```
Traceback (most recent call last):
  File "/mnt/c/Users/pchekin/intel/intel-xpu-backend-for-triton/scripts/./compare-ci-runs/compare_runs.py", line 302, in <module>
    main()
  File "/mnt/c/Users/pchekin/intel/intel-xpu-backend-for-triton/scripts/./compare-ci-runs/compare_runs.py", line 276, in main
    df = parse_directory(args.triton_benchmark, num_cfg, None, num_dir)
  File "/mnt/c/Users/pchekin/intel/intel-xpu-backend-for-triton/scripts/./compare-ci-runs/compare_runs.py", line 130, in parse_directory
    df = parse_triton_benchmark_data(config, df, file) if triton_benchmark else parse_pytorch_benchmark_data(
  File "/mnt/c/Users/pchekin/intel/intel-xpu-backend-for-triton/scripts/./compare-ci-runs/compare_runs.py", line 112, in parse_triton_benchmark_data
    raw_data = pd.read_csv(file, header=0, usecols=shape + ["Triton-TFlops", "XeTLA-TFlops"])
  File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 1026, in read_csv
    return _read(filepath_or_buffer, kwds)
  File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 620, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 1620, in __init__
    self._engine = self._make_engine(f, self.engine)
  File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 1898, in _make_engine
    return mapping[engine](f, **self.options)
  File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/c_parser_wrapper.py", line 140, in __init__
    self._validate_usecols_names(usecols, self.orig_names)
  File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/base_parser.py", line 979, in _validate_usecols_names
    raise ValueError(
ValueError: Usecols do not match columns, columns expected but not found: ['XeTLA-TFlops']
```

**Comment by ESI-SYD** [comment]
> compare-ci-runs/compare_runs.py --triton-benchmark -n dev:10752862880 -d ref:10752866039

`XeTLA` GEMM data missed in your two runs

**Comment by ESI-SYD** [comment]
> @ESI-SYD I've tried the script and it didn't work for me:
> 
> ```
> ./compare-ci-runs/compare_runs.py --triton-benchmark -n dev:10752862880 -d ref:10752866039
> ```
> 
> ```
> Traceback (most recent call last):
>   File "/mnt/c/Users/pchekin/intel/intel-xpu-backend-for-triton/scripts/./compare-ci-runs/compare_runs.py", line 302, in <module>
>     main()
>   File "/mnt/c/Users/pchekin/intel/intel-xpu-backend-for-triton/scripts/./compare-ci-runs/compare_runs.py", line 276, in main
>     df = parse_directory(args.triton_benchmark, num_cfg, None, num_dir)
>   File "/mnt/c/Users/pchekin/intel/intel-xpu-backend-for-triton/scripts/./compare-ci-runs/compare_runs.py", line 130, in parse_directory
>     df = parse_triton_benchmark_data(config, df, file) if triton_benchmark else parse_pytorch_benchmark_data(
>   File "/mnt/c/Users/pchekin/intel/intel-xpu-backend-for-triton/scripts/./compare-ci-runs/compare_runs.py", line 112, in parse_triton_benchmark_data
>     raw_data = pd.read_csv(file, header=0, usecols=shape + ["Triton-TFlops", "XeTLA-TFlops"])
>   File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 1026, in read_csv
>     return _read(filepath_or_buffer, kwds)
>   File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 620, in _read
>     parser = TextFileReader(filepath_or_buffer, **kwds)
>   File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 1620, in __init__
>     self._engine = self._make_engine(f, self.engine)
>   File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 1898, in _make_engine
>     return mapping[engine](f, **self.options)
>   File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/c_parser_wrapper.py", line 140, in __init__
>     self._validate_usecols_names(usecols, self.orig_names)
>   File "/home/paul/venv/triton-3.10/lib/python3.10/site-packages/pandas/io/parsers/base_parser.py", line 979, in _validate_usecols_names
>     raise ValueError(
> ValueError: Usecols do not match columns, columns expected but not found: ['XeTLA-TFlops']
> ```

Fixed in https://github.com/intel/intel-xpu-backend-for-triton/pull/2060/commits/ccdd64158386c383622432cc52b7ac729f8ace35

**Comment by ESI-SYD** [comment]
@sommerlukas @pbchekin Can you take another look for this PR? Thanks

**Comment by victor-eds** [comment]
> @sommerlukas @pbchekin Can you take another look for this PR? Thanks

Lukas is off. I'll take on :+1:

File: scripts/compare-ci-runs/compare_runs.py [hunk 4]
```diff
@@ -97,25 +96,65 @@ def parse_data(config: str, df: pd.DataFrame, file: Path) -> pd.DataFrame:
...
+        xetla_raw_data.rename(columns={"tflops": f"XeTLA-TFlops-{config}"}, inplace=True)
+        return triton_raw_data.merge(xetla_raw_data, how="outer", on=["params", "benchmark"])
+    except FileNotFoundError:
+        print(f"Warning: One or both files not found: {triton_file} or {xetla_file}")
+        return pd.DataFrame()
+
+
+def build_triton_benchmark_reports_path(directory: Path, report_name: str) -> str:
+    """Construct the full file path for a given report name."""
+    return os.path.join(directory, f"benchmark-reports/{report_name}-report.csv")
```
Comment (victor-eds) [inline scripts/compare-ci-runs/compare_runs.py:116]:
```suggestion
def build_triton_benchmark_reports_path(directory: Path, report_name: str) -> str:
    """Construct the full file path for a given report name."""
    return os.path.join(directory, "benchmark-reports", f"{report_name}-report.csv")
```
NIT: use `os.path.join`

File: scripts/compare-ci-runs/compare_runs.py [hunk 5]
```diff
@@ -97,25 +96,65 @@ def parse_data(config: str, df: pd.DataFrame, file: Path) -> pd.DataFrame:
...
+    reports_files = {
+        "softmax": ("softmax-triton", "softmax-xetla"), "gemm": ("gemm-triton", "gemm-xetla"), "attn":
+        ("attn-triton", "attn-xetla")
+    }
+
+    reports_list = [df]
+    for (triton_file, xetla_file) in reports_files.values():
+        triton_path = build_triton_benchmark_reports_path(Path(directory), triton_file)
+        xetla_path = build_triton_benchmark_reports_path(Path(directory), xetla_file)
+        reports_list.append(merge_triton_xetla_reports_data(config, triton_path, xetla_path))
```
Comment (victor-eds) [inline scripts/compare-ci-runs/compare_runs.py:133]:
```suggestion
    reports = ["softmax", "gemm", "attn"]

    reports_list = [df]
    for report in reports:
        triton_file = f"{report}-triton"
        xetla_file = f"{report}-xetla"
        triton_path = build_triton_benchmark_reports_path(Path(directory), triton_file)
        xetla_path = build_triton_benchmark_reports_path(Path(directory), xetla_file)
        reports_list.append(merge_triton_xetla_reports_data(config, triton_path, xetla_path))
```

File: scripts/compare-ci-runs/compare_runs.py [hunk 6]
```diff
@@ -97,25 +96,65 @@ def parse_data(config: str, df: pd.DataFrame, file: Path) -> pd.DataFrame:
...
+        df = parse_triton_benchmark_data(config, df, directory)
+    else:
+        df = pd.DataFrame(columns=["dev", "name", "batch_size", f"speedup {config}", "suite", "datatype", "mode"])
+        for file in Path(directory).rglob("*performance.csv"):
+            df = parse_pytorch_benchmark_data(config, df, file)
 
     if previous is not None:
-        df = df.merge(previous, how="outer", on=["suite", "datatype", "mode", "name", "dev"])
+        df = df.merge(previous, how="outer", on=["params", "benchmark"]) if triton_benchmark else df.merge(
+            previous, how="outer", on=["suite", "datatype", "mode", "name", "dev"])
```
Comment (victor-eds) [inline scripts/compare-ci-runs/compare_runs.py:151]:
```suggestion
        on = ["params", "benchmark"] if triton_benchmark else ["suite", "datatype", "mode", "name", "dev"]
        df = df.merge(previous, how="outer", on=on)
```

More readable IMO

File: scripts/compare-ci-runs/compare_runs.py [hunk 7]
```diff
@@ -169,12 +208,13 @@ def eval_data(df: pd.DataFrame, numerator: str, denominator: str, plot: bool):
         import matplotlib.pyplot as plt
         from matplotlib.backends.backend_pdf import PdfPages
 
-        df["xlabel"] = df[["suite", "mode", "datatype"]].agg(", ".join, axis=1)
+        df["xlabel"] = df[["params", "benchmark"]].agg(
+            ", ".join, axis=1) if triton_benchmark else df[["suite", "mode", "datatype"]].agg(", ".join, axis=1)
```
Comment (victor-eds) [inline scripts/compare-ci-runs/compare_runs.py:212]:
```suggestion
        keys = ["params", "benchmark"] if triton_benchmark else ["suite", "mode", "datatype"]
        df["xlabel"] = df[keys].agg(", ".join, axis=1)
```
Readability

**Review by victor-eds — APPROVED** [review]
Not a big expert, but I don't see any major issues. LGTM as is

File: scripts/compare-ci-runs/compare_runs.py [hunk 8]
```diff
@@ -97,25 +96,64 @@ def parse_data(config: str, df: pd.DataFrame, file: Path) -> pd.DataFrame:
...
+        print(f"Warning: One or both files not found: {triton_file} or {xetla_file}")
+        return pd.DataFrame()
+
+
+def build_triton_benchmark_reports_path(directory: Path, report_name: str) -> str:
+    """Construct the full file path for a given report name."""
+    return os.path.join(directory, "benchmark-reports", f"{report_name}-report.csv")
+
+
+def parse_triton_benchmark_data(config: str, df: pd.DataFrame, directory: Path) -> pd.DataFrame:
```
Comment (anmyachev) [inline scripts/compare-ci-runs/compare_runs.py:119]:
```suggestion
def parse_triton_benchmark_data(config: str, directory: Path) -> pd.DataFrame:
```

File: scripts/compare-ci-runs/compare_runs.py [hunk 9]
```diff
@@ -97,25 +96,64 @@ def parse_data(config: str, df: pd.DataFrame, file: Path) -> pd.DataFrame:
...
+    """Parse triton benchmark data from a merged dataframe into the dataframe.
+        Now focus on dft path for softmax, gemm and attention
+        which include both xetla and triton data with regular name."""
+
+    reports = ["softmax", "gemm", "attn"]
+
+    reports_list = [df]
```
Comment (anmyachev) [inline scripts/compare-ci-runs/compare_runs.py:126]:
```suggestion
    reports_list = []
```

File: scripts/compare-ci-runs/compare_runs.py [hunk 10]
```diff
@@ -97,25 +96,64 @@ def parse_data(config: str, df: pd.DataFrame, file: Path) -> pd.DataFrame:
...
+    for report in reports:
+        triton_file = f"{report}-triton"
+        xetla_file = f"{report}-xetla"
+        triton_path = build_triton_benchmark_reports_path(Path(directory), triton_file)
+        xetla_path = build_triton_benchmark_reports_path(Path(directory), xetla_file)
```
Comment (anmyachev) [inline scripts/compare-ci-runs/compare_runs.py:131]:
```suggestion
        triton_path = build_triton_benchmark_reports_path(directory, triton_file)
        xetla_path = build_triton_benchmark_reports_path(directory, xetla_file)
```

File: scripts/compare-ci-runs/compare_runs.py [hunk 11]
```diff
@@ -97,25 +96,64 @@ def parse_data(config: str, df: pd.DataFrame, file: Path) -> pd.DataFrame:
...
+
+def parse_directory(triton_benchmark: bool, config: str, previous: pd.DataFrame, directory: Path) -> pd.DataFrame:
     """Parse all CSV files for a configuration in a directory, merging with
         the previous dataframe if present."""
-    df = pd.DataFrame(columns=["dev", "name", "batch_size", f"speedup {config}", "suite", "datatype", "mode"])
-    for file in Path(directory).rglob("*performance.csv"):
-        df = parse_data(config, df, file)
+    if triton_benchmark:
+        df = pd.DataFrame()
+        df = parse_triton_benchmark_data(config, df, directory)
```
Comment (anmyachev) [inline scripts/compare-ci-runs/compare_runs.py:142]:
```suggestion
        df = parse_triton_benchmark_data(config, directory)
```

File: scripts/compare-ci-runs/compare_runs.py [hunk 12]
```diff
@@ -97,25 +96,64 @@ def parse_data(config: str, df: pd.DataFrame, file: Path) -> pd.DataFrame:
...
+        triton_file = f"{report}-triton"
+        xetla_file = f"{report}-xetla"
+        triton_path = build_triton_benchmark_reports_path(Path(directory), triton_file)
+        xetla_path = build_triton_benchmark_reports_path(Path(directory), xetla_file)
+        reports_list.append(merge_triton_xetla_reports_data(config, triton_path, xetla_path))
+
+    return pd.concat(reports_list, ignore_index=True)
+
+
+def parse_directory(triton_benchmark: bool, config: str, previous: pd.DataFrame, directory: Path) -> pd.DataFrame:
```
Comment (anmyachev) [inline scripts/compare-ci-runs/compare_runs.py:137]:
```suggestion
def parse_directory(triton_benchmark: bool, config: str, directory: Path) -> pd.DataFrame:
```

File: scripts/compare-ci-runs/compare_runs.py [hunk 13]
```diff
@@ -230,19 +288,29 @@ def main():
             print("Failed to obtain raw data")
             sys.exit(1)
 
-        df = parse_directory(num_cfg, None, num_dir)
-        df = parse_directory(denom_cfg, df, denom_dir)
+        df = parse_directory(args.triton_benchmark, num_cfg, None, num_dir)
+        df = parse_directory(args.triton_benchmark, denom_cfg, df, denom_dir)
```
Comment (anmyachev) [inline scripts/compare-ci-runs/compare_runs.py:292]:
```suggestion
        num_df = parse_directory(args.triton_benchmark, num_cfg, num_dir)
        denom_df = parse_directory(args.triton_benchmark, denom_cfg, denom_dir)

        on = ["params", "benchmark"] if args.triton_benchmark else ["suite", "datatype", "mode", "name", "dev"]
        df = denom_df.merge(num_df, how="outer", on=on)
```

File: scripts/compare-ci-runs/compare_runs.py (hunk 11 above)
Comment (anmyachev) [reply scripts/compare-ci-runs/compare_runs.py:142]: Parameter `df` appears to be unused, looks safe to remove

File: scripts/compare-ci-runs/compare_runs.py (hunk 10 above)
Comment (anmyachev) [reply scripts/compare-ci-runs/compare_runs.py:131]: According python typings `directory` is already of `Path` type, there is no need to construct it again.

File: scripts/compare-ci-runs/compare_runs.py (hunk 13 above)
Comment (anmyachev) [reply scripts/compare-ci-runs/compare_runs.py:292]: The merge looks redundant in `parse_directory` function, by moving it out the code becomes more readable.

Comment (ESI-SYD) [reply scripts/compare-ci-runs/compare_runs.py]: Done: https://github.com/intel/intel-xpu-backend-for-triton/pull/2060/commits/4a5f8bdf86f15bac85eb17c1017a781fb757838c

**Review by anmyachev — APPROVED** [review]

**Pull request merged**